src/verilog/verilog_preprocessor.cpp

```diff
@@ -155,43 +214,94 @@ Function: verilog_preprocessort::replace_macros
 
 \*******************************************************************/
 
-void verilog_preprocessort::replace_macros(std::string &s)
+auto verilog_preprocessort::parse_define_parameters() -> definet::parameterst
 {
-  std::string dest;
+  tokenizer().next_token(); // eat (
 
-  for(unsigned i=0; i<s.size();)
-  {
-    if(s[i]=='`')
-    {
-      i++;
-      unsigned start=i;
+  definet::parameterst result;
 
-      while(i<s.size() && 
-            (isalnum(s[i]) || s[i]=='$' || s[i]=='_'))
-        i++;
+  while(true)
+  {
+    tokenizer().skip_ws();
 
-      std::string text(s, start, i-start);
+    auto parameter = tokenizer().next_token();
+    if(!parameter.is_identifier())
+      throw verilog_preprocessor_errort() << "expecting a define parameter";
 
-      definest::const_iterator it=defines.find(text);
+    result.push_back(parameter.text);
 
-      if(it==defines.end())
-      {
-        error() << "unknown preprocessor macro \"" << text << "\"" << eom;
-        throw 0;
-      }
+    tokenizer().skip_ws();
 
-      // found it! replace it!
+    auto token = tokenizer().next_token();
 
-      dest+=it->second;
-    }
+    if(token == ')')
+      break; // done
+    else if(token == ',')
+      continue;           // keep going
+    else if(token == '=') // SystemVerilog 2009
+      throw verilog_preprocessor_errort()
+        << "default parameters are not supported yet";
     else
+      throw verilog_preprocessor_errort() << "expecting a define parameter";
+  }
+
+  return result;
+}
+
+/*******************************************************************\
+
+Function: verilog_preprocessort::parse_define_arguments
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+auto verilog_preprocessort::parse_define_arguments(const definet &define)
+  -> std::map<std::string, std::vector<tokent>>
+{
+  if(define.parameters.empty())
+    return {};
+
+  if(tokenizer().next_token() != '(')
+    throw verilog_preprocessor_errort() << "expecting define arguments";
+
+  std::vector<std::vector<tokent>> arguments = {{}};
+
```

Comment:
Unnecessary right-hand side?

Reply:
Oh, no! The RHS creates an element in the outer vector, which contains an empty vector. This is important.
I'll add a comment.